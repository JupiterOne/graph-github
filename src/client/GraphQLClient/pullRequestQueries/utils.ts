import { URL } from 'url';

const innerResourcePaginationRequired = (pullRequest): boolean => {
  if (!pullRequest) {
    return false;
  }
  return (
    pullRequest.commits?.pageInfo.hasNextPage ||
    pullRequest.reviews?.pageInfo.hasNextPage ||
    pullRequest.labels?.pageInfo.hasNextPage
  );
};

const responseToResource = (node) => {
  return {
    ...node,
    commits: node.commits?.nodes?.map((node) => node.commit) ?? [],
    reviews: node.reviews?.nodes ?? [],
    labels: node.labels?.nodes ?? [],
  };
};

const hasRepoOwnerAndName = (node) => {
  const repoOwnerAndName = determineRepoOwnerAndName(node);
  return repoOwnerAndName.repoOwner && repoOwnerAndName.repoName;
};

const determineRepoOwnerAndName = (
  node,
): { repoOwner?: string; repoName?: string } => {
  const urlPath = node.url // ex: https://github.com/JupiterOne/graph-github/pull/1
    ? new URL(node.url)?.pathname // ex: /JupiterOne/graph-github/pull/4"
    : '';

  // Attempt to pull repo name and owner from graphQL response. If not there, parse the pull request url.
  const repoOwner = node.headRepository?.owner?.login ?? urlPath.split('/')[1]; // ex: JupiterOne
  const repoName = node.headRepository?.name ?? urlPath.split('/')[2]; // ex: graph-github

  return {
    repoOwner,
    repoName,
  };
};

export default {
  innerResourcePaginationRequired,
  responseToResource,
  findRepoOwnerAndName: determineRepoOwnerAndName,
  hasRepoOwnerAndName,
};
