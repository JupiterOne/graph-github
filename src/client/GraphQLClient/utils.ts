/* eslint-disable @typescript-eslint/no-duplicate-enum-values */
import { gte } from 'semver';
import { first } from 'lodash';

const responseToResource = (node) => {
  if (!hasProperties(node)) {
    return null;
  }

  const associatedPullRequest = first(
    node.mergeCommit?.associatedPullRequests?.nodes,
  );

  delete node.mergeCommit?.associatedPullRequests;

  return {
    ...node,
    ...(node.mergeCommit && {
      mergeCommit: {
        ...node.mergeCommit,
        associatedPullRequest,
      },
    }),
    ...(node.commits && {
      commits:
        node.commits.nodes?.filter((node) => node).map((node) => node.commit) ??
        [],
    }),
    ...(node.reviews && {
      reviews: node.reviews.nodes?.filter((node) => node) ?? [],
    }),
    ...(node.labels && {
      labels: node.labels.nodes?.filter((node) => node) ?? [],
    }),
  };
};

const hasProperties = (object: any) => {
  try {
    return Object.keys(object).length > 0;
  } catch (error) {
    return false;
  }
};

/**
 * Specifies the version a feature was added to GHE Server.
 */
export enum EnterpriseFeatures {
  REPO_VULN_ALERT_STATE_ARG = '3.5.0',
  REPO_VULN_ALERT_FIELDS = '3.5.0', // fixedAt, number, state were added
  BRANCH_PROTECTION_RULES_BLOCKS_CREATIONS_FIELD = '3.5.0',
  BRANCH_PROTECTION_RULES_APP_MEMBER = '3.6.0',
  /**
   * Docs link: https://docs.github.com/en/enterprise-server@3.5/rest/code-scanning#list-code-scanning-alerts-for-an-organization
   */
  LIST_CODE_SCANNING_ALERT_FOR_ORG = '3.5.0',
  /**
   * Docs link: https://docs.github.com/en/enterprise-server@3.5/rest/secret-scanning/secret-scanning#list-secret-scanning-alerts-for-an-organization
   */
  LIST_SECRET_SCANNING_ALERT_FOR_ORG = '3.5.0',
}

/**
 * Returns t/f if the provided feature is supported for the provided server version.
 * @param featureVersion
 * @param gheServerVersion
 */
const isSupported = (
  featureVersion: EnterpriseFeatures,
  gheServerVersion?: string | null,
): boolean => {
  if (!gheServerVersion || gheServerVersion.length === 0) {
    // All features are supported if on GH Cloud.
    return true;
  }

  return gte(gheServerVersion, featureVersion);
};

export default {
  hasProperties,
  responseToResource,
  isSupported,
};
