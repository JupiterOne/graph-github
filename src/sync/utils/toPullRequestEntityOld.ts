import {
  parseTimePropertyValue,
  setRawData,
} from '@jupiterone/integration-sdk-core';
import { omit, uniq } from 'lodash';
import { GITHUB_PR_ENTITY_TYPE, GITHUB_PR_ENTITY_CLASS } from '../../constants';
import {
  PullsListResponseItem,
  PullsListCommitsResponseItem,
  IdEntityMap,
  UserEntity,
  PullRequestEntity,
} from '../../types';
import {
  flattenMatrix,
  aggregateProperties,
  displayNamesFromLogins,
} from '../../util/propertyHelpers';
import { Approval } from '../converters';

export function toPullRequestEntityOld(
  data: PullsListResponseItem,
  commits?: PullsListCommitsResponseItem[],
  commitsApproved?: PullsListCommitsResponseItem[],
  commitsByUnknownAuthor?: PullsListCommitsResponseItem[],
  approvals?: Approval[],
  usersByLogin?: IdEntityMap<UserEntity>,
): PullRequestEntity {
  const commitHashes = commits ? commits.map((c) => c.sha) : undefined;
  const commitMessages = commits
    ? commits.map((c) => c.commit.message)
    : undefined;
  const commitsApprovedHashes = commitsApproved
    ? commitsApproved.map((c) => c.sha)
    : undefined;
  const commitsByUnknownAuthorHashes = commitsByUnknownAuthor
    ? commitsByUnknownAuthor.map((c) => c.sha)
    : undefined;
  const commitsNotApproved = commitHashes
    ? commitHashes.filter((c) => !commitsApprovedHashes!.includes(c))
    : undefined;

  const approved = commitsNotApproved
    ? commitsNotApproved.length === 0
    : undefined;
  const validated = commitsByUnknownAuthor
    ? commitsByUnknownAuthor.length === 0
    : undefined;

  const approverLogins = approvals
    ? uniq(
        flattenMatrix<string>(
          aggregateProperties<string[]>('approverUsernames', approvals),
        ),
      )
    : undefined;
  const userLogin: string = data.user ? data.user.login : '';
  const authorUser = (usersByLogin || {})[userLogin];
  let reviewerLogins: string[] = [];
  let reviewers: string[] = [];
  if (data.requested_reviewers) {
    reviewerLogins = aggregateProperties<string>(
      'login',
      data.requested_reviewers,
    );
    reviewers = data.requested_reviewers.reduce(
      (reviewers: string[], reviewerData) => {
        if (reviewerData) {
          if (usersByLogin && usersByLogin[reviewerData.login]) {
            reviewers.push(usersByLogin[reviewerData.login].displayName!);
          } else {
            reviewers.push(reviewerData.login);
          }
        }
        return reviewers;
      },
      [],
    );
  }
  const entity: PullRequestEntity = {
    _type: GITHUB_PR_ENTITY_TYPE,
    _class: [GITHUB_PR_ENTITY_CLASS],
    _key: `${data.base.repo.full_name}/pull-requests/${data.number}`,
    displayName: `${data.base.repo.name}/${data.number}`,
    accountLogin: data.base.repo.owner ? data.base.repo.owner.login : '',
    repository: data.base.repo.name,
    // the type is hacked here because typing of data properties is controlled by a library call
    // so I can't just say that data.number is a string
    // here would be another way to solve it:
    // id: JSON.stringify(data.number).replace(/\"/g, ''),
    id: <string>(<unknown>data.number),

    name: data.title,
    title: data.title,
    description:
      data.body && data.body.length > 0
        ? `${data.body.substring(0, 80)}...`
        : undefined,
    webLink: data.html_url,

    state: data.state,
    open: data.state === 'open',
    mergeCommitHash: data.merge_commit_sha,
    merged: (data.merged_at as any) !== null,
    declined: data.state === 'closed' && (data.merged_at as any) === null,
    approved,
    validated,

    commits: commitHashes,
    commitMessages,
    commitsApproved: commitsApprovedHashes,
    commitsNotApproved,
    commitsByUnknownAuthor: commitsByUnknownAuthorHashes,

    source: data.head.ref,
    target: data.base.ref,

    createdOn: parseTimePropertyValue(data.created_at),
    updatedOn: parseTimePropertyValue(data.updated_at),
    mergedOn: parseTimePropertyValue(data.merged_at),

    authorLogin: userLogin,
    author: authorUser
      ? // We know displayName is set; see toOrganizationMemberEntity
        authorUser.displayName!
      : // Fallback to username. This will always be used when ingesting from a
        // user account, since we don't ingest team members in that case (there
        // is no team)
        userLogin,
    reviewerLogins: reviewerLogins,
    reviewers: reviewers,
    approverLogins,
    approvers:
      approverLogins && usersByLogin
        ? displayNamesFromLogins(approverLogins, usersByLogin)
        : approverLogins,
  };
  const rawDataPropertiesToRemove = ['head', 'base']; // a few particularly large pieces of data that are repeated on every PR
  setRawData(entity, {
    name: 'default',
    rawData: omit(data, rawDataPropertiesToRemove),
  });
  return entity;
}
