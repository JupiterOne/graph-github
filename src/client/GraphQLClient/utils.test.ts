import utils, { EnterpriseFeatures } from './utils';

describe('utils', () => {
  test('hasProperties', () => {
    expect(utils.hasProperties(null)).toBeFalsy();
    expect(utils.hasProperties(undefined)).toBeFalsy();
    expect(utils.hasProperties({})).toBeFalsy();
    expect(utils.hasProperties(new Object())).toBeFalsy();
    expect(utils.hasProperties({ dog: 'dog' })).toBeTruthy();
  });

  test('responseToResource', () => {
    expect(
      utils.responseToResource({
        id: 'PR_asdf',
        commits: {
          totalCount: 11,
          nodes: [null, null, null, null, null, { commit: { asdf: 'commit' } }],
          pageInfo: {
            endCursor: 'MTE',
            hasNextPage: false,
          },
        },
        reviews: {
          totalCount: 117,
          nodes: [
            null,
            undefined,
            {
              id: 'asdf',
              commit: null,
              author: {},
              state: 'COMMENTED',
              submittedAt: '2000-04-19T04:33:08Z',
              updatedAt: '2000-04-19T04:33:08Z',
              url: 'https://github.com/',
            },
          ],
          pageInfo: {
            endCursor: 'asdf333',
            hasNextPage: true,
          },
        },
        labels: {
          totalCount: 2,
          nodes: [null, { asdf: 'label' }],
          pageInfo: {
            endCursor: null,
            hasNextPage: false,
          },
        },
      }),
    ).toMatchSnapshot();

    expect(
      utils.responseToResource({
        mergeCommit: {
          id: 'C_kwDOIAVVA9oAKDg1NjNlOWIyZDYzMzRiNzUwZjhmN2I2MzA4NTgxODg5NTBkMmFiOTI',
          associatedPullRequests: {
            nodes: [
              {
                number: 2,
                id: 'PR_kwDOIAVVA84_DshD',
                state: 'MERGED',
                reviewDecision: null,
                url: 'https://github.com/j1-ingest/ubiquitous-umbrella/pull/2',
              },
            ],
          },
          author: {
            user: {
              login: 'VDubber',
            },
          },
        },
      }),
    ).toMatchSnapshot();
  });
  test('#isSupported', () => {
    expect(
      utils.isSupported(
        EnterpriseFeatures.REPO_VULN_ALERT_STATE_ARG,
        '3.3.300',
      ),
    ).toBeFalsy();
    expect(
      utils.isSupported(EnterpriseFeatures.REPO_VULN_ALERT_FIELDS, '3.5.0'),
    ).toBeTruthy();
    expect(
      utils.isSupported(EnterpriseFeatures.REPO_VULN_ALERT_FIELDS, '3.6.3'),
    ).toBeTruthy();
    expect(
      utils.isSupported(EnterpriseFeatures.REPO_VULN_ALERT_FIELDS, ''),
    ).toBeTruthy();
    expect(
      utils.isSupported(EnterpriseFeatures.REPO_VULN_ALERT_FIELDS, null),
    ).toBeTruthy();
  });
});
