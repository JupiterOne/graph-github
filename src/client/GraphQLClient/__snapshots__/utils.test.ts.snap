// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils responseToResource 1`] = `
Object {
  "commits": Array [
    Object {
      "asdf": "commit",
    },
  ],
  "id": "PR_asdf",
  "labels": Array [
    Object {
      "asdf": "label",
    },
  ],
  "reviews": Array [
    Object {
      "author": Object {},
      "commit": null,
      "id": "asdf",
      "state": "COMMENTED",
      "submittedAt": "2000-04-19T04:33:08Z",
      "updatedAt": "2000-04-19T04:33:08Z",
      "url": "https://github.com/",
    },
  ],
}
`;

exports[`utils responseToResource 2`] = `
Object {
  "mergeCommit": Object {
    "associatedPullRequest": Object {
      "id": "PR_kwDOIAVVA84_DshD",
      "number": 2,
      "reviewDecision": null,
      "state": "MERGED",
      "url": "https://github.com/j1-ingest/ubiquitous-umbrella/pull/2",
    },
    "author": Object {
      "user": Object {
        "login": "VDubber",
      },
    },
    "id": "C_kwDOIAVVA9oAKDg1NjNlOWIyZDYzMzRiNzUwZjhmN2I2MzA4NTgxODg5NTBkMmFiOTI",
  },
}
`;
